File: 백준/Gold/16236. 아기 상어/아기 상어.java

import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

//		물고기 한마리씩 잡아 먹으면서
//		갈 수 있는 모든 곳을 pq에 넣음
//		pq에서 뽑아내는데, 조건에 해당하면 먹을 위치 뱉어줌

		n = sc.nextInt();
		grid = new int[n][n];

		int r = -1;
		int c = -1; // 상어 위치

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				grid[i][j] = sc.nextInt();
				if (grid[i][j] == 9) {
					r = i;
					c = j;
					grid[i][j] = 0; // 걸리적 거리지 않게 위치만 저장하고 0 처리
				}
			}
		}

		shark = 2;
		int eat = 0; // 상어 크기만큼 먹으면 커질거임
		int time = 0;
		while (true) {
			int[] fish = find(r, c); // 이동할 위치(물고기냠)
			if (fish == null) { // 이동 못하면 끝!
				break;
			}

			r = fish[0];
			c = fish[1];
			time += fish[2];
			eat++;
			if (eat == shark) {
				shark++; // 상어 커지기
				eat = 0; 
			}
		}
		System.out.println(time);

	}

	static int n;
	static int shark;
	static int[] row = { -1, 1, 0, 0 };
	static int[] col = { 0, 0, 1, -1 };
	static int[][] grid;

	private static int[] find(int r, int c) {
		boolean[][] visited = new boolean[n][n];
		visited[r][c] = true;
		PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<>() {
			@Override
			public int compare(int[] o1, int[] o2) { // 뱉어낼 조건 설정
				if (o1[2] == o2[2]) {
					if (o1[0] == o2[0]) {
						return Integer.compare(o1[1], o2[1]); // 3 r도 같으면 c 작은 순
					}
					return Integer.compare(o1[0], o2[0]); // 2. 거리 같으면 r 작은 순
				}

				return Integer.compare(o1[2], o2[2]); // 1. 거리순
			}
		});
		pq.add(new int[] { r, c, 0 }); // 초기 상어 위치 넣어주기

		while (!pq.isEmpty()) {
			int size = pq.size();
			for (int i = 0; i < size; i++) {
				int[] node = pq.poll();
				int x = node[0];
				int y = node[1];
				int d = node[2];
				if (grid[x][y] > 0 && grid[x][y] < shark) { // 뱉어낼 조건 해당하면 뱉어내기(알아서 우선순위에 맞게 뱉어낼거임)
					grid[x][y] = 0;
					return new int[] { x, y, d };
				}
				for (int k = 0; k < 4; k++) {
					int nr = x + row[k];
					int nc = y + col[k];
					if (nr >= 0 && nr < n && nc >= 0 && nc < n && !visited[nr][nc] && grid[nr][nc] <= shark) { // 이동할 수 있는 위치 다 pq에 넣음
						visited[nr][nc] = true; // pq에 다 들어가도 잡아먹을 수 있는 애를 뱉어낼거기 때문에 이동할 수 있는 곳 다 넣어도 상관 없음.
						pq.add(new int[] { nr, nc, d + 1 });
					}
				}
			}

		}

		return null;
	}
}

File: 백준/Gold/16236. 아기 상어/아기 상어.py
'''
문제설명
    초기 전투로봇 레벨2
    1초에 상하좌우 한칸식 이동
    자기보다 레벨 높은애들은 지나갈 수 없음
    자기보다 레벨 낮은애들 없앨 수 있음
    자기랑 레벨 같으면 지나갈 수 는 있음
    가는 우선순위
    1. 거리가 가장 가까운
    2. 가장 위쪽, 왼쪽
    없으면 끝
    전투로봇은 본인 레벨만큼 먹으면 레벨 상승함

입력
    9: 전투로봇
    1~6 : 몬스터
구상
    bfs도는데 힙큐 사용.

'''
import heapq
from collections import deque

n = int(input())
grid = [list(map(int, input().split())) for i in range(n)]

r, c, level, eat = -1, -1, 2, 0
for i in range(n):
    for j in range(n):
        if grid[i][j] == 9:
            grid[i][j] = 0
            r, c = i, j

time = 0
row = [-1, 1, 0, 0]
col = [0, 0, 1, -1]
while True:
    # 자기보다 레벨 높은애들은 지나갈 수 없음
    # 자기보다 레벨 낮은애들 없앨 수 있음
    # 자기랑 레벨 같으면 지나갈 수 는 있음
    target_q = []

    # 아 거리 계산하는게 쉽지 않네...

    q = deque([(r, c, 0)])
    visited = [[False] * n for i in range(n)]
    visited[r][c] = True
    while q:
        r, c, dist = q.popleft()

        for k in range(4):
            nr = r + row[k]
            nc = c + col[k]
            if not (0 <= nr < n and 0 <= nc < n) or visited[nr][nc]:
                continue
            if grid[nr][nc] == 0 or grid[nr][nc] == level:
                visited[nr][nc] = True
                q.append((nr, nc, dist + 1))
            elif grid[nr][nc] < level:
                visited[nr][nc] = True
                heapq.heappush(target_q, (dist + 1, nr, nc))
    if not target_q:
        break
    # print(target_q)
    dist, target_r, target_c = heapq.heappop(target_q)
    # print((target_r,target_c),dist)
    grid[target_r][target_c] = 0  # 먹는다
    eat += 1
    r = target_r
    c = target_c
    time += dist
    if eat == level:
        level += 1
        eat = 0

print(time)
